tests/PHPStan/Rules/Classes/data/bug-5333.php

```diff
@@ -0,0 +1,80 @@
+<?php declare(strict_types = 1);
+
+namespace Bug5333;
+
+use function PHPStan\Testing\assertNativeType;
+use function PHPStan\Testing\assertType;
+
+class Route {}
+
+class HelloWorld
+{
+	/**
+	 * @param Route|Route[] $foo
+	 *
+	 * @return Route
+	 **/
+	public function sayHello($foo): Route
+	{
+		if (\is_array($foo)) {
+			$res = $foo[0];
+			assertType('Bug5333\Route', $res);
+			assertNativeType('mixed', $res);
+
+			if (!$res instanceof Route) {
+				throw new \Exception();
+			}
+
+			assertType('array<Bug5333\Route>', $foo);
+			assertNativeType('array', $foo);
+
+			assertType('Bug5333\Route', $res);
+			assertNativeType('Bug5333\Route', $res);
+
+			return $res;
+		}
+
+		return $foo;
+	}
+}
+
+class HelloWorld2
+{
+	/**
+	 * @var Route|callable():Route
+	 **/
+	private $foo;
+
+	/**
+	 * @param Route|callable():Route $foo
+	 **/
+	public function setFoo($foo): void
+	{
+		assertType('Bug5333\Route|(callable(): Bug5333\Route)', $foo);
+		assertNativeType('mixed', $foo);
+
+		$this->foo = $foo;
+	}
+
+	public function getFoo(): Route
+	{
+		assertType('Bug5333\Route|(callable(): Bug5333\Route)', $this->foo);
+		assertNativeType('Bug5333\Route|(callable(): Bug5333\Route)', $this->foo);
+
+		if (\is_callable($this->foo)) {
+			assertType('(Bug5333\Route&callable(): mixed)|(callable(): Bug5333\Route)', $this->foo);
+			assertNativeType('(Bug5333\Route&callable(): mixed)|(callable(): Bug5333\Route)', $this->foo);
+
+			$res = ($this->foo)();
+			assertType('mixed', $res);
+			assertNativeType('mixed', $res);
```

Comment:
can we be more precise here?

```suggestion
			assertType('callable(): Bug5333\Route', $this->foo);
			assertNativeType('(Bug5333\Route&callable(): mixed)|(callable(): Bug5333\Route)', $this->foo);

			$res = ($this->foo)();
			assertType('Bug5333\Route', $res);
			assertNativeType('mixed', $res);
```

Reply:
Please add a third example and use FinalRoute (final class) instead of Route. You'll see it's gonna be more precise.